src/core/src/mapping.rs

```diff
@@ -65,8 +66,72 @@ impl StdError for Error {
     }
 }
 
-fn lock_mapping<R>(access: memory::Access, buffer: &buffer::Raw<R>)
-                   -> Result<MutexGuard<R::Mapping>, Error>
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct Raw<R: Resources> {
+    resource: UnsafeCell<R::Mapping>,
+    accessible: AtomicBool,
+}
+
+#[doc(hidden)]
+impl<R: Resources> Raw<R> {
+    pub fn new(resource: R::Mapping) -> Self {
+        Raw {
+            resource: UnsafeCell::new(resource),
+            accessible: AtomicBool::new(true),
+        }
+    }
+
+    pub unsafe fn take_access(&self) -> bool {
+        self.accessible.swap(false, atomic::Ordering::Relaxed)
+    }
+
+    pub unsafe fn release_access(&self) {
+        self.accessible.store(true, atomic::Ordering::Relaxed)
+    }
+
+    pub unsafe fn use_access(&self) -> &mut R::Mapping {
+        &mut *self.resource.get()
```

Comment:
we should verify that `accessible == true`, using a debug assertion to avoid performance hit